Align behavior with objects raising in `__getattr__`

perrinjerome opened the pull request:
The observed problem was a behavior different between C and python
implementation on python 3, happening with Zope python script. When the
context can not be accessed by the current user, Zope binds a
`Shared.DC.Scripts.Bindings.UnauthorizedBinding`, a class that raises an
Unauthorized error when the context is actually accessed, in order to
postpone the Unauthorized if something is actually accessed. This class
does implements this by raising Unauthorized in `__getattr__`.

The python implementation of `rolesForPermissionOn` used `hasattr` and
`hasattr` has changed between python2 and python3, on python2 it was
ignoring all exceptions, including potential Unauthorized errors and
just returning False, but on python3 these errors are now raised.
This change of behavior of python causes `rolesForPermissionOn` to
behave differently: when using python implementation on python2 or when
using C implementation, such Unauthorized errors were gracefully handled
and caused `checkPermission` to return False, but on python3 the
Unauthorized is raised.

The C implementation of `rolesForPermissionOn` uses a construct
equivalent to the python2 version of `hasattr`. For consistency - and
because ignoring errors is usually not good - we also want to change it
to be have like the python3 implementation.

This change make this scenario behave the same between python and C
implementations:
- `Unauthorized` errors raised in `__getattr__` are supported on py3.
- Other errors  than `AttributeError` and `Unauthorized` raised in
     `__getattr__` are no longer ignored in the C implementation.

perrinjerome commented:
Some more notes:
 - `guarded_getattr` also has the same logic of explicitly tolerating `Unauthorized` errors https://github.com/zopefoundation/AccessControl/blob/27d88c40e251b370f4dd2fcc7ae03c2967c68e4c/src/AccessControl/ZopeGuards.py#L68-L73
 - The test reproduces a simplified version of the problem I described, it happens when calling a zope python script with proxy role, when running with security manager with "ownerous" set to false. By defining `__ac_local_roles__ = {}` like in the test class, it uses the same code path.
 - In production the problem I observed is probably a rare case, but it seemed OK to make the code python3 implementation behaves same as python2 and C regarding this.

dataflake reviewed (approved):
I can't claim to understand the implications 100% but it looks like a sensible change.

d-maurer commented on src/AccessControl/cAccessControl.c:
```diff
@@ -1847,13 +1847,16 @@ c_rolesForPermissionOn(PyObject *perm, PyObject *object,
+        else:
       */
       PyObject *roles = PyObject_GetAttr(object, n);
       if (roles != NULL)
```
This still gives us a difference for the Python 3 and C implementations: while the Python 3 implementation only ignores `AttributeError` and `Unauthorized`, the C implementation ignores all exceptions (like Python 2 did).

I am not sure whether this has a practical effect but I think the Python 3 and C implementation should behave the same way (for consistency).

d-maurer replied:
We should change the C implementation: ignoring all exceptions may misinterpret temporary exceptions (like e.g. `DatabaseReadConflict`).

perrinjerome replied:
Thanks, I was also starting to think the same. This still matches the python2 behavior of `hasattr`, but the python3 behavior is better.

I'll try to take care of this (but maybe not right now)

perrinjerome replied:
C implementation was changed as discussed here

perrinjerome commented:
I have updated the changes here to make the `C` implementation behave same as the python3 implementation: only `AttributeError` and `Unauthorized` exceptions are allowed when getting the permission attribute on objects and other exceptions are re-raised. I have also adjusted the commit message of the commit changing the behavior of `__getattr__` to describe the behavior.

For the testing, I have changed the test of `PermissionRole` to test both the python and C implementations and implemented some more low level tests there. This is another seperate preparatory commit.

d-maurer commented on CHANGES.rst:
```diff
@@ -8,6 +8,9 @@ For changes before version 3.0, see ``HISTORY.rst``.
 
 - Respect ``PURE_PYTHON`` environment variable set to ``0`` when running tests.
 
+- Make Python implementation behave same as C implementation regarding
+  objects raising exceptions in ``__getattr__``.
```
```suggestion
- Let the roles access in ``rolesForPermissionOn`` interpret ``AttributeError`` and ``Unauthorized`` as "no roles definition for this permission at this object" and report any other exception (for the Python and C implementation). We have to treat ``Unauthorized`` like ``AttributeError`` to support ``Shared.DC.Scripts.Bindings.UnauthorizedBinding`` which raises ``Unauthorized`` for any access.
```

perrinjerome replied:
Thank you very much, this is a great description. I'm integrating this and merging.
